examples/votchallenge/main.rs

```diff
@@ -0,0 +1,164 @@
+mod trax_protocol;
+
+use std::io::stdin;
+
+use mosse::{MosseTrackerSettings, MultiMosseTracker};
+
+use crate::trax_protocol::{
+    ChannelType, Image, ImageType, Region, RegionType, TraxMessageFromClient, TraxMessageFromServer,
+};
+
+#[derive(Debug)]
+pub enum ServerState {
+    Introduction,
+    Initialization,
+    Reporting {
+        multi_tracker: MultiMosseTracker,
+        first_region: Region,
+    },
+    Termination,
+}
+
+struct MosseTraxServer {
+    state: ServerState,
+}
+impl Default for MosseTraxServer {
+    fn default() -> Self {
+        Self {
+            state: ServerState::Introduction,
+        }
+    }
+}
+impl MosseTraxServer {
+    fn run(mut self) {
+        log::info!("starting run");
+
+        println!("{}", self.make_hello_message());
+
+        for line in stdin().lines() {
+            let line = line.unwrap();
+            log::trace!("handling line: {line:?}");
+            let message: TraxMessageFromClient = line.parse().unwrap();
+            let response = self.process_message(message);
+            println!("{}", response);
+        }
+    }
+
+    fn make_hello_message(&mut self) -> TraxMessageFromServer {
+        TraxMessageFromServer::Hello {
+            version: 1,
+            name: "MosseRust".to_string(),
+            identifier: "mosse-tracker-rust".to_string(),
+            image: ImageType::Path,
+            region: RegionType::Rectangle,
+            channels: vec![ChannelType::Color],
+        }
+    }
+
+    fn process_message(&mut self, message: TraxMessageFromClient) -> TraxMessageFromServer {
+        match message {
+            TraxMessageFromClient::Initialize { image, region } => self.process_init(image, region),
+            TraxMessageFromClient::Frame { images } => self.process_frame(images),
+            // FIXME: return Result from this function, and make the outer loop print "quit" and exit on error?
+            TraxMessageFromClient::Quit => panic!("client sent quit message"),
+        }
+    }
+
+    fn process_init(&mut self, image: Image, region: Region) -> TraxMessageFromServer {
+        let first = image.open().unwrap();
+
+        // initialize a new model
+        let (width, height) = first.to_rgb8().dimensions();
+        // FIXME: This tracks a square that entirely encloses the target region, so it may fixate
+        // on the background for tall or wide targets.
+        let window_size = f64::max(region.width, region.height) as u32; // size of the tracking window
```

Comment:
I tried changing this a few times, and added it to the git commit messages

max(region.width, region.height) gives us scores of
    EAO: 0.070
    Accuracy: 0.355
    Robustness: 0.195

min(region.width, region.height) gets:
    EAO: 0.064
    Accuracy: 0.369
    Robustness: 0.166

128x128 pixel square gets:
    EAO: 0.090
    Accuracy: 0.373
    Robustness: 0.264

These numbers are among the bottom of the field that the VOT challenge benchmark evaluated, but I think that's okay, because this tracker is mostly built for speed rather than state of the art tracking performance. 

For reference, a null tracker (which just returns the original region every time) would score:
EAO: 0.048
Accuracy: 0.304
Robustness: 0.146

Reply:
Nice overview! Results make sense though. There's quite a bit of room for improvement, maybe even from a correctness point of view.

WRT 'built for speed': is the perf (FPS?) of the other algorithms posted anywhere? I never really expected this implementation to actually be faster than the competition, as I only implemented a single pass of trivial optimisations. Would be fun if it were the case, though.

Reply:
> WRT 'built for speed': is the perf (FPS?) of the other algorithms posted anywhere?

From the results report, a lot of the ones that performed well on VOT-ST failed to meet the "real time" requirement (I'm guessing 30/60fps or something?) and therefore scored poorly on the VOT-RT test the first year they introduced it, but then the second year people had tuned their algorithms a bit, and the best entrants were also capable of realtime tracking. I cant remember where I read this though.

In other news, I tried running it on an older dataset and everything trucks along fine until send me this:

```
@@TRAX:initialize "file:///Users/alsuren/src/mosse-tracker/examples/votchallenge/sequences/godfather/color/00000203.jpg" "340282346638528859811704183484516925440.0000,340282346638528859811704183484516925440.0000,-inf,-inf" 
```

which we understandably crap out on. I tried hacking around it by treating that `initialize` message as a `frame` message, but we end up with a bit of a nonsense html report.

The next person who wants to come in and improve the mosse-tracker library will probably want to do it on a sequence-by-sequence basis, and check the results by hand.

If you enable the debug image output (search for `.predicted.png` main.rs) and then:

```bash
vot test --sequence=sequences/bolt1 MosseRust
```

then you can debug a frame at a time.

I also submitted an issue upstream in case they want to take my rust implementation of the protocol and make it official (https://github.com/votchallenge/toolkit/issues/67). Let's see how that goes.

I think I'm going to stop working on this now. Hopefully this is useful enough as it is.